module-bsp/board/rt1051/puretx/hal/battery_charger/BatteryCharger.cpp

```diff
@@ -1,120 +1,201 @@
+        case Status::ChargingDone:
+            return NewState::ChargingDone;
+        case Status::ChargerNotCharging:
+            return NewState ::Discharging;
+        default:
+            return NewState::PluggedNotCharging;
+        }
     }
 
-    void BatteryCharger::BatteryCharger::deinit()
-    {
-        bsp::battery_charger::deinit();
-    }
+    /// A few constants to make code readability better
+    constexpr auto int_b_soc_change =
+        static_cast<std::uint16_t>(bsp::battery_charger::batteryINTBSource::SOCOnePercentChange);
+    constexpr auto int_b_max_temp = static_cast<std::uint16_t>(bsp::battery_charger::batteryINTBSource::maxTemp);
+    constexpr auto int_b_min_temp = static_cast<std::uint16_t>(bsp::battery_charger::batteryINTBSource::minTemp);
+    constexpr auto int_b_all      = static_cast<std::uint16_t>(bsp::battery_charger::batteryINTBSource::all);
+    constexpr auto int_b_miv_v    = static_cast<std::uint16_t>(bsp::battery_charger::batteryINTBSource::minVAlert);
+
+    constexpr auto int_source_charger =
+        static_cast<std::uint8_t>(bsp::battery_charger::topControllerIRQsource::CHGR_INT);
+
+    constexpr auto int_source_fuel_gauge =
+        static_cast<std::uint8_t>(bsp::battery_charger::topControllerIRQsource::FG_INT);
 
-    void BatteryCharger::processStateChangeNotification(std::uint8_t notification)
+} // namespace
+
+namespace hal::battery
+{
+    class PureBatteryCharger : public AbstractBatteryCharger
     {
-        if (notification == static_cast<std::uint8_t>(bsp::battery_charger::batteryIRQSource::INTB) ||
-            notification == static_cast<std::uint8_t>(bsp::battery_charger::batteryIRQSource::INOKB)) {
-            checkBatteryChargerInterrupts();
-        }
+      public:
+        struct IrqEvents
+        {
+            enum class Type
+            {
+                Controller,
+                USBChargerAttached
+            };
+            Type type;
+            std::uint8_t chargerType{};
+        };
+        using BatteryWorkerQueue = WorkerQueue<IrqEvents>;
+
+        explicit PureBatteryCharger(xQueueHandle irqQueueHandle);
+        ~PureBatteryCharger();
+
+        Voltage getBatteryVoltage() const final;
+        SOC getSOC() const final;
+        ChargingStatus getChargingStatus() const final;
+
+        static BatteryWorkerQueue &getWorkerQueueHandle();
+
+      private:
+        static constexpr auto workerStackSize = 1024;
+        void setChargingCurrentLimit(std::uint8_t usbType);
+        void sendNotification(AbstractBatteryCharger::Events event);
+        void checkControllerInterrupts();
+
+        QueueHandle_t notificationChannel;
+        inline static std::unique_ptr<BatteryWorkerQueue> workerQueue;
+    };
+
+    PureBatteryCharger::PureBatteryCharger(QueueHandle_t irqQueueHandle) : notificationChannel{irqQueueHandle}
+    {
+        workerQueue = std::make_unique<BatteryWorkerQueue>(
+            "battery_charger",
+            [this](const auto &msg) {
```

Comment:
tihs might actually be an issue in cases when there would be further handling based on freertos timer and callback? ( callback to not existing element)